sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java

```diff
@@ -362,8 +366,11 @@ private void createRunnerAndConsumersForPTransformRecursively(
       return BeamFnApi.InstructionResponse.newBuilder().setProcessBundle(response);
     } catch (Exception e) {
       // Make sure we clean-up from the active set of bundle processors.
+      BeamFnLoggingMDC.setInstructionId("");
       bundleProcessorCache.discard(bundleProcessor);
       throw e;
+    } finally {
+      BeamFnLoggingMDC.setInstructionId("");
```

Comment:
```suggestion
      bundleProcessorCache.discard(bundleProcessor);
      throw e;
    } finally {
      BeamFnLoggingMDC.setInstructionId(null);
```

The finally is always executed even when an exception is caught.

Reply:
Applied suggestion. But I had a concern, which is why I didn't initially do this.

I don't think this it is true that finally will be called in all cases.
Such as if an exception is thrown from the catch block (which we do here), and the thrown exception is not handled further up the stack. (Though I suppose we would crash the program entirely? Or could the thread somehow revive).

https://stackoverflow.com/questions/1555567/when-is-finally-run-if-you-throw-an-exception-from-the-catch-block/1555578


Am I wrong? Do we need to be concerned about this scenario?